cmd/notation/plugin/install.go

```diff
@@ -0,0 +1,329 @@
+}
+
+// installPluginFromTarGz extracts and untar a plugin tar.gz file, validates and
+// installs the plugin
+func installPluginFromTarGz(ctx context.Context, tarGzPath string, force bool) error {
+	logger := log.GetLogger(ctx)
+	rc, err := os.Open(tarGzPath)
+	if err != nil {
+		return err
+	}
+	defer rc.Close()
+	decompressedStream, err := gzip.NewReader(rc)
+	if err != nil {
+		return err
+	}
+	defer decompressedStream.Close()
+	tarReader := tar.NewReader(decompressedStream)
+	// require one and only one file with name in the format
+	// `notation-{plugin-name}`
+	for {
+		header, err := tarReader.Next()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+		if !header.FileInfo().Mode().IsRegular() || strings.Contains(header.Name, "..") {
+			continue
+		}
+		// validate and get plugin name from file name
+		pluginName, err := notationplugin.ExtractPluginNameFromFileName(header.Name)
+		if err != nil {
+			logger.Infof("processing file %s. File name not in format notation-{plugin-name}, skipped", header.Name)
+			continue
+		}
+		return installPluginExecutable(ctx, header.Name, pluginName, tarReader, force)
+	}
+	return fmt.Errorf("no valid plugin file was found in %s. Plugin file name must in format notation-{plugin-name}", tarGzPath)
+}
+
+// installPluginExecutable extracts, validates, and installs a plugin file
+func installPluginExecutable(ctx context.Context, fileName string, pluginName string, fileReader io.Reader, force bool) error {
+	// sanity check
+	if runtime.GOOS == "windows" && filepath.Ext(fileName) != ".exe" {
+		return fmt.Errorf("on Windows, plugin executable file %s is missing the '.exe' extension", fileName)
+	}
+	if runtime.GOOS != "windows" && filepath.Ext(fileName) == ".exe" {
+		return fmt.Errorf("on %s, plugin executable file %s cannot have the '.exe' extension", runtime.GOOS, fileName)
+	}
+	// extract to tmp dir
+	tmpDir, err := os.MkdirTemp(".", notationPluginTmpDir)
+	if err != nil {
+		return fmt.Errorf("failed to create notationPluginTmpDir: %w", err)
+	}
+	defer os.RemoveAll(tmpDir)
+	tmpFilePath := filepath.Join(tmpDir, fileName)
+	pluginFile, err := os.OpenFile(tmpFilePath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0700)
+	if err != nil {
+		return err
+	}
+	if _, err := io.Copy(pluginFile, fileReader); err != nil {
+		defer pluginFile.Close()
+		return err
+	}
+	if err := pluginFile.Close(); err != nil {
+		return err
+	}
+	// get plugin metadata
+	pluginMetadata, err := notationplugin.GetPluginMetadata(ctx, pluginName, tmpFilePath)
+	if err != nil {
+		return err
+	}
+	pluginVersion := pluginMetadata.Version
+	// check plugin existence and version
+	var currentPluginVersion string
+	currentPluginMetadata, err := notationplugin.GetPluginMetadataIfExist(ctx, pluginName)
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return err
+	}
+	if err == nil { // plugin already exists
```

Comment:
nit: it is more readable and less error prone.
```suggestion
	if err != nil {
		if !errors.Is(err, os.ErrNotExist) {
			return err
		}
	} else { // plugin already exists
```